### PR TITLE
Remove and indirection for document id, for less memory footprint, an…

### PR DESCRIPTION
**Pull request by baldersheim**
…d better generated code.

@vekterli PR

### CODE REVIEW

File: document/src/vespa/document/base/idstring.h [hunk 1]
```diff
@@ -13,108 +12,60 @@ namespace document {
...
-    virtual vespalib::stringref getNamespaceSpecific() const = 0;
-    virtual LocationType getLocation() const = 0;
-    virtual std::pair<int16_t, int64_t> getGidBitsOverride() const { return std::pair<int16_t, int64_t>(0, 0); }
-    virtual bool hasDocType() const { return false; }
-    virtual vespalib::stringref getDocType() const { return ""; }
-    virtual bool hasNumber() const { return false; }
-    virtual uint64_t getNumber() const { return 0; }
-    virtual bool hasGroup() const { return false; }
-    virtual vespalib::stringref getGroup() const { return ""; }
+    bool hasDocType() const { return size(1 != 0); }
```
Comment (vekterli) [inline document/src/vespa/document/base/idstring.h]: Was this intended to be `return size(1) != 0;`? Funnily enough, due to implicit type conversions this ends up being the same expression in practice

File: document/src/vespa/document/base/idstring.h [hunk 2]
```diff
@@ -13,108 +12,60 @@ namespace document {
...
-        Offsets(uint32_t maxComponents, uint32_t first, vespalib::stringref id);
-        uint16_t first() const { return _offsets[0]; }
+        Offsets() = default;
+        uint16_t compute(vespalib::stringref id);
         uint16_t operator [] (size_t i) const { return _offsets[i]; }
-        size_t numComponents() const { return _numComponents; }
+        static const Offsets DefaultID;
     private:
+        Offsets(vespalib::stringref id);
         uint16_t _offsets[5];
```
Comment (vekterli) [inline document/src/vespa/document/base/idstring.h]: Consider exposing and using `MAX_COMPONENTS` here to reduce number of magical constants

File: document/src/vespa/document/base/idstring.cpp [hunk 3]
```diff
@@ -219,21 +170,62 @@ void setLocation(IdString::LocationType &loc, IdString::LocationType val,
...
+      _location(0),
+      _offsets(),
       _groupOffset(0),
       _has_number(false)
 {
     // TODO(magnarn): Require that keys are lexicographically ordered.
-    validate();
+    verifyIdString(id.data(), id.size());
+    validate(_offsets.compute(id));
+
```
Comment (vekterli) [inline document/src/vespa/document/base/idstring.cpp]: Consider removing extra linebreak

**Review by vekterli — COMMENTED** [review]
See question on `hasDocType`, otherwise LGTM

File: document/src/vespa/document/base/idstring.h [hunk 4]
```diff
@@ -13,108 +12,60 @@ namespace document {
  * \class document::IdString
  * \ingroup base
  *
- * \brief Superclass for all document identifier schemes.
+ * \brief New scheme for documents with no forced distribution.
+ *
+ * By using this scheme, documents will be evenly distributed within VDS,
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.h]: Cleaned up.

File: document/src/vespa/document/base/idstring.h (hunk 1 above)
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.h]: Good catch. this was a nasty little bugger. That was what I thought I wrote. At the test started working again :)

File: document/src/vespa/document/base/idstring.h (hunk 2 above)
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.h]: Done.

File: document/src/vespa/document/base/idstring.cpp (hunk 3 above)
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.cpp]: Fixed

File: document/src/vespa/document/base/idstring.h [hunk 5]
```diff
@@ -13,108 +12,60 @@ namespace document {
...
 
-    const vespalib::string & toString() const;
+    const vespalib::string & toString() const { return _rawId; }
 
-protected:
-    IdString(uint32_t maxComponents, uint32_t namespaceOffset, vespalib::stringref rawId);
+private:
     size_t offset(size_t index) const { return _offsets[index]; }
-    size_t size(size_t index) const { return _offsets[index+1] - _offsets[index] - 1; }
+    size_t size(size_t index) const { return std::max(0, _offsets[index+1] - _offsets[index] - 1); }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.h]: Done

File: document/src/vespa/document/bucket/bucketidfactory.cpp [hunk 6]
```diff
@@ -42,29 +42,16 @@ BucketIdFactory::getBucketId(const DocumentId& id) const
 {
     uint64_t location = id.getScheme().getLocation();
     assert(GlobalId::LENGTH >= sizeof(uint64_t) + 4u);
-    uint64_t gid = reinterpret_cast<const uint64_t&>(
-            *(id.getGlobalId().get() + 4));
+    uint64_t gid = reinterpret_cast<const uint64_t&>(*(id.getGlobalId().get() + 4));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/bucket/bucketidfactory.cpp]: Done.

**Pull request merged**